src/common/src/hash/key.rs

```diff
@@ -60,7 +60,7 @@ impl HashCode {
 
 pub trait HashKeySerializer {
     type K: HashKey;
-    fn from_hash_code(hash_code: HashCode) -> Self;
+    fn from_hash_code(hash_code: HashCode, estimated_key_size: usize) -> Self;
```

Comment:
Also we can pass the `key_len: usize here which means the number of hash key fields here so that the `null_bitmap` can reserve memory too.

Reply:
> Also we can pass the key_len: usize here which means the number of hash key fields here so that the null_bitmap` can reserve memory too.

I tried this previously, but performance did not really improve (either there were only ~1% improvements, or even regressions). I suspect extra allocation is probably the cost, since not every row has null values.

Will use the approach suggested here: https://github.com/risingwavelabs/risingwave/issues/8802#issuecomment-1486390916 to optimize `null_bitmap` in a different PR.

```
vec ser varchar KeySerialized, 128 rows, Pr[null]=0
                        time:   [22.043 µs 22.070 µs 22.098 µs]
                        change: [-2.0727% -1.7386% -1.4289%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

vec ser varchar KeySerialized, 1024 rows, Pr[null]=0
                        time:   [177.50 µs 178.02 µs 178.57 µs]
                        change: [-1.1887% -0.9216% -0.6614%] (p = 0.00 < 0.05)
                        Change within noise threshold.

vec ser varchar KeySerialized, 128 rows, Pr[null]=0.01
                        time:   [21.983 µs 21.998 µs 22.013 µs]
                        change: [-1.1509% -0.9599% -0.7895%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

vec ser varchar KeySerialized, 1024 rows, Pr[null]=0.01
                        time:   [159.93 µs 160.51 µs 161.12 µs]
                        change: [+0.3612% +0.7191% +1.0570%] (p = 0.00 < 0.05)
                        Change within noise threshold.

vec ser varchar KeySerialized, 128 rows, Pr[null]=0.1
                        time:   [17.316 µs 17.337 µs 17.360 µs]
                        change: [-1.8434% -1.6448% -1.4519%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

vec ser varchar KeySerialized, 1024 rows, Pr[null]=0.1
                        time:   [143.92 µs 144.47 µs 145.05 µs]
                        change: [+2.3449% +2.7724% +3.2537%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```